### PR TITLE
[AKS] `az aks`: Support IP-based Load Balancer and Instance Level Public IP Tags

### PR DESCRIPTION
**Pull request by nilo19**
**Related command**

```
az aks create --load-balancer-backend-pool-type {nodeIPConfiguration|nodeIP}
az aks update --load-balancer-backend-pool-type {nodeIPConfiguration|nodeIP}
az aks create --enable-node-public-ip --node-public-ip-tags
az aks nodepool add --enable-node-public-ip --node-public-ip-tags
```

**Description**

Support the above commands. Ref: https://github.com/Azure/azure-cli-extensions/pull/5536 and https://github.com/Azure/azure-cli-extensions/pull/5355

**Testing Guide**

```
# create a cluster with load balancer backend pool type
az aks create --load-balancer-backend-pool-type {nodeIPConfiguration|nodeIP}
# update a cluster with load balancer backend pool type
az aks update --load-balancer-backend-pool-type {nodeIPConfiguration|nodeIP}
# create a cluster with node public ip tags
az aks create --enable-node-public-ip --node-public-ip-tags RoutingPreference=Internet
# create an agent pool with node public ip tags
az aks nodepool add --enable-node-public-ip --node-public-ip-tags RoutingPreference=Internet
```

**History Notes**

[AKS] `az aks create/update`: Add new parameter `--load-balancer-backend-pool-type` to define the Load Balancer backend pool type of managed inbound backend pool
[AKS] `az aks create`: Add parameter `--node-public-ip-tags` to specify the ipTags of the node public IPs
[AKS] `az aks nodepool`: Add parameter `--node-public-ip-tags` to specify the ipTags of the node public IPs

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

File: src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py [hunk 1]
```diff
@@ -1601,6 +1615,14 @@ def add_agentpool(self, agentpool: AgentPool) -> AgentPool:
             headers=self.context.get_aks_custom_headers(),
         )
 
+    def set_up_agentpool_network_profile(self, agentpool: AgentPool) -> AgentPool:
+        self._ensure_agentpool(agentpool)
+        ip_tags = self.context.get_ip_tags()
+        if ip_tags:
+            agentpool.network_profile = self.models.AgentPoolNetworkProfile()
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py:1622]: this might remove existing configs

Comment (nilo19) [reply src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py:1622]: yes, but I need https://github.com/Azure/azure-cli/pull/27900 to get merged first before I can fix this.

Comment (nilo19) [reply src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py]: fixed

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py [hunk 2]
```diff
@@ -9253,3 +9253,94 @@ def test_aks_update_upgrade_settings(self, resource_group, resource_group_locati
...
+    #
+    #     # update
+    #     update_cmd = 'aks update -g {resource_group} -n {name} --load-balancer-backend-pool-type=nodeIPConfiguration'
+    #     self.cmd(update_cmd, checks=[
+    #         self.check('networkProfile.loadBalancerProfile.backendPoolType', 'nodeIPConfiguration'),
+    #     ])
+
+    @AllowLargeResponse()
+    @AKSCustomResourceGroupPreparer(random_name_length=17, name_prefix='clitest', location='westus2')
+    def test_node_public_ip_tags(self, resource_group, resource_group_location):
```
[the remark this answers is not in the document]
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:9579]: Test passed

**Comment by FumingZhang** [comment]
The PR title is a little messy, would you split it into several notes and put it under section `History Notes` in PR description?

**Comment by nilo19** [comment]
> The PR title is a little messy, would you split it into several notes and put it under section `History Notes` in PR description?

Changed. Would this be acceptable?

**Comment by FumingZhang** [comment]
please keep the formatting, taking `az aks create --load-balancer-backend-pool-type {nodeIPConfiguration|nodeIP}` as an example,

- [AKS] `az aks create`: Add option `--load-balancer-backend-pool-type`

would be better

File: src/azure-cli/azure/cli/command_modules/acs/_help.py [hunk 3]
```diff
@@ -180,6 +180,10 @@
     type: int
     short-summary: Load balancer idle timeout in minutes.
     long-summary: Desired idle timeout for load balancer outbound flows, default is 30 minutes. Please specify a value in the range of [4, 100].
+  - name: --load-balancer-backend-pool-type
+    type: string
+    short-summary: Load balancer backend pool type.
+    long-summary: Load balancer backend pool type, supported values are nodeIP and nodeIPConfiguration.
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/acs/_help.py:186]: If the value of this parameter is enumerable, it is suggested to use `arg_type=get_enum_type()` to define the param. After that, it is not necessary to manually specify the available options.

Comment (nilo19) [reply src/azure-cli/azure/cli/command_modules/acs/_help.py]: Fixed, please take a look.

File: src/azure-cli/azure/cli/command_modules/acs/_params.py [hunk 4]
```diff
@@ -217,6 +226,8 @@ def load_arguments(self, _):
         c.argument('load_balancer_outbound_ip_prefixes', validator=validate_load_balancer_outbound_ip_prefixes)
         c.argument('load_balancer_outbound_ports', type=int, validator=validate_load_balancer_outbound_ports)
         c.argument('load_balancer_idle_timeout', type=int, validator=validate_load_balancer_idle_timeout)
+        c.argument('load_balancer_backend_pool_type', arg_type=get_enum_type(backend_pool_types),
+                   validator=validate_load_balancer_backend_pool_type)
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/acs/_params.py:230]: After defined with `arg_type=get_enum_type(backend_pool_types)`, it is not necessary to validate the input manually.

Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/acs/_params.py:230]: After defined with `arg_type=get_enum_type(backend_pool_types)`, it is not necessary to validate the input manually.
```suggestion
        c.argument('load_balancer_backend_pool_type', arg_type=get_enum_type(backend_pool_types))
```

Comment (nilo19) [reply src/azure-cli/azure/cli/command_modules/acs/_params.py]: removed

**Comment by nilo19** [comment]
@FumingZhang can we get this merged?

**Review by yanzhudd — APPROVED** [review]

**Comment by nilo19** [comment]
@jsntcy @kairu-ms @zhoxing-ms can we merge this today?

**Comment by yanzhudd** [comment]
> @jsntcy @kairu-ms @zhoxing-ms can we merge this today?

Sorry, PRs cannot be merged during our release process. Once the release is complete, we would merge it.

File: src/azure-cli/azure/cli/command_modules/acs/_help.py [hunk 5]
```diff
@@ -664,6 +671,10 @@
     type: int
     short-summary: Load balancer idle timeout in minutes.
     long-summary: Desired idle timeout for load balancer outbound flows, default is 30 minutes. Please specify a value in the range of [4, 100].
+  - name: --load-balancer-backend-pool-type
+    type: string
+    short-summary: Load balancer backend pool type.
+    long-summary: Defines the LoadBalancer backend pool type of managed inbound backend pool. nodeIP means the VMs will be attached to the LoadBalancer by adding its private IP address to the backend pool. nodeIPConfiguration means the VMs will be attached to the LoadBalancer by referencing the backend pool ID in the VM's NIC.
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/acs/_help.py:677]:
```suggestion
    long-summary: Define the LoadBalancer backend pool type of managed inbound backend pool. The nodeIP means the VMs will be attached to the LoadBalancer by adding its private IP address to the backend pool. nodeIPConfiguration means the VMs will be attached to the LoadBalancer by referencing the backend pool ID in the VM's NIC.
```

File: src/azure-cli/azure/cli/command_modules/acs/_help.py [hunk 6]
```diff
@@ -180,6 +180,10 @@
     type: int
     short-summary: Load balancer idle timeout in minutes.
     long-summary: Desired idle timeout for load balancer outbound flows, default is 30 minutes. Please specify a value in the range of [4, 100].
+  - name: --load-balancer-backend-pool-type
+    type: string
+    short-summary: Load balancer backend pool type.
+    long-summary: Define the LoadBalancer backend pool type of managed inbound backend pool. nodeIP means the VMs will be attached to the LoadBalancer by adding its private IP address to the backend pool. nodeIPConfiguration means the VMs will be attached to the LoadBalancer by referencing the backend pool ID in the VM's NIC.
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/acs/_help.py:186]:
```suggestion
    long-summary: Define the LoadBalancer backend pool type of managed inbound backend pool. The nodeIP means the VMs will be attached to the LoadBalancer by adding its private IP address to the backend pool. The nodeIPConfiguration means the VMs will be attached to the LoadBalancer by referencing the backend pool ID in the VM's NIC.
```

**Comment by nilo19** [comment]
@zhoxing-ms any other comments?

**Review by zhoxing-ms — APPROVED** [review]

**Review by wangzelin007 — APPROVED** [review]

**Pull request merged**